enhanced-stack-trace-queries.md

```diff
@@ -0,0 +1,111 @@
+# Enhanced stack trace queries
+
+## Backround
+
+The SDKs have a built-in `__stack_trace` query that can be used display where the workflow is blocked at the time of the
+query (the last processed workflow task).
+
+The response to a `__stack_trace` query is a single string representing a list of stack individual stack traces
+separated by double newlines.
+
+Example from the TS SDK:
+
+```
+    at someInternalFunction (/path/to/workflows/my-workflow.ts:18:10)
+    at someWorkflowFunction (/path/to/workflows/my-workflow.ts:60:20)
+
+    at someInternalFunction2 (/path/to/workflows/my-workflow.ts:100:30)
+    at someWorkflowFunction2 (/path/to/workflows/my-workflow.ts:123:5)
+```
+
+While this is valuable for developers to be able to understand how their code correlates to event histories, it is quite
+limited since it requires looking up the actual code to get the full picture.
+
+## Proposal
+
+SDKs can optionally implement an `__enhanced_stack_trace` query (name TBD) where instead of returning the stack traces
+as a single string, they return a pre-defined structure, that structure will also include file sources.
+
+The web UI can use this response to show the workflow code and highlight locations where it is blocked.
+
+### Proposed structure
+
+```ts
+export interface SDKInfo {
+  name: string;
+  version: string;
+}
+
+/**
+ * Represents a slice of a file starting at lineOffset
+ */
+export interface FileSlice {
+  content: string;
```

Comment:
Given that the worker will anyway have to do something like a split in order to extract the range of lines to be sent, then either returning an array of lines or normalizing the EOL sequence to a single \n seems like a trivial task.

Reply:
I also note that once serialized to JSON, an array of line has essentially the same length as a normalized multiline string (EOL become `\\n` after JSON serialization).

Reply:
> seems like a trivial task.

Definitely trivial, but without a comment saying that's what's expected, nobody will know to do it. I don't mind whichever way is chosen so long as it's made known.

Reply:
I'll mention that content lines must end with `\n` and that it's the SDK's responsibility

Reply:
Thinking about this some more, blindly sending chunks to the UI will hurt the ability to have proper syntax highlighting.
I think we should keep this structure for future optimization but I'd start with sending entire files for now and make the UI deal with presenting snippets instead of entire files.

Reply:
A large Java trace for example may be sending over dozens of large files here. This is all starting to make me rethink sending sources is that important. I definitely won't use it as a user, but then again I haven't seen all of the complaints from users that can't see their code in the UI (if they even exist).

How do other "APM" or error tracking systems do this? So when Sentry shows code for an error, what is it doing? Does it show code for every element in the trace or only the top element of the trace? How do they obtain the source code to map to the lines? Is it gathered like we'd be doing here when you invoke Sentry "report" or whatever? I'll try to dig into https://github.com/getsentry/sentry-go a bit.

Maybe we should only send the file of the top element of the trace?

Reply:
Updated comment